src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -1095,117 +1093,331 @@ impl Tree {
+                    .build()
+                })
+        };
+
+        let visitor = TreeVisitor { nodes, loc: self };
+        match visit_children {
+            ChildrenVisitMode::VisitChildrenOfAccessed =>
+                visitor.traverse_this_parents_children_other(
+                    access_source,
+                    node_skipper,
+                    node_app,
+                )?,
+            ChildrenVisitMode::SkipChildrenOfAccessed =>
+                visitor.traverse_nonchildren(access_source, node_skipper, node_app)?,
+        }
+        interp_ok(root.into_inner().unwrap())
+    }
+
+    /// Performs a wildcard access on the tree with root `root`. Takes the `access_relatedness`
+    /// for each node from the `WildcardState` datastructure.
+    /// * `root`: Root of the tree being accessed.
+    /// * `access_source`: the index of the accessed tag, if any.
+    ///   This is only used for printing the correct tag on errors.
+    /// * `max_local_tag`: For any tag larger than this we do not perform local accesses. If no
+    ///   foreign access to a node with a larger tag is possible then we return an access error.
+    fn perform_wildcard_access(
+        &mut self,
+        root: UniIndex,
+        access_source: Option<UniIndex>,
+        max_local_tag: Option<BorTag>,
+        nodes: &mut UniValMap<Node>,
+        loc_range: Range<u64>,
+        access_range: Option<AllocRange>,
+        access_kind: AccessKind,
+        access_cause: diagnostics::AccessCause,
+        global: &GlobalState,
+        alloc_id: AllocId, // diagnostics
+        span: Span,        // diagnostics
+    ) -> InterpResult<'tcx> {
+        // This does a traversal starting from the root through the tree updating
+        // the permissions of each node.
+        // The difference to `perform_access` is that we take the access
+        // relatedness from the wildcard tracking state of the node instead of
+        // from the visitor itself.
+        TreeVisitor { loc: self, nodes }
+            .traverse_this_parents_children_other(
+                root,
+                |args: &NodeAppArgs<'_>| -> ContinueTraversal {
+                    let node = args.nodes.get(args.idx).unwrap();
+                    let perm = args.loc.perms.get(args.idx);
+                    let wildcard_state =
+                        args.loc.wildcard_accesses.get(args.idx).cloned().unwrap_or_default();
+
+                    let only_foreign = max_local_tag
+                        .map(|max_local_tag| max_local_tag < node.tag)
+                        .unwrap_or(false);
+                    let old_state = perm.copied().unwrap_or_else(|| node.default_location_state());
+                    // If we know where, relative to this node, the wildcard access occurs,
+                    // then check if we can skip the entire subtree.
+                    if let Some(relatedness) =
+                        wildcard_state.access_relatedness(access_kind, only_foreign)
```

Comment:
Can we avoid duplicating the access_relatedness code by moving it into a local closure?